frontend/include/chpl/types/CPtrType.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_TYPES_CPTR_TYPE_H
+#define CHPL_TYPES_CPTR_TYPE_H
+
+#include "chpl/types/Type.h"
+#include "chpl/types/QualifiedType.h"
+
+namespace chpl {
+namespace types {
+
+class CPtrType final : public Type {
+ private:
+  const CPtrType* instantiatedFrom_;
+  QualifiedType eltType_;
+
+  CPtrType(const CPtrType* instantiatedFrom, QualifiedType eltType)
+    : Type(typetags::CPtrType), instantiatedFrom_(instantiatedFrom),
+      eltType_(std::move(eltType)) {
+    // not an instantiation -> eltType_ should be empty
+    CHPL_ASSERT(instantiatedFrom_ != nullptr || eltType_ == QualifiedType());
+    // is an instantiation -> eltType should not be empty
+    CHPL_ASSERT(instantiatedFrom_ == nullptr || eltType_ != QualifiedType());
+  }
+
+  bool contentsMatchInner(const Type* other) const override {
+    auto rhs = (CPtrType*) other;
+    return instantiatedFrom_ == rhs->instantiatedFrom_ &&
+           eltType_ == rhs->eltType_;
+  }
+
+  void markUniqueStringsInner(Context* context) const override {}
+
+  Genericity genericity() const override {
+    return MAYBE_GENERIC;
```

Comment:
Seems easy enough to be more accurate with this one. If there is no eltType, it's generic. Otherwise, it's generic iff the eltType is (so you could return `eltTypeTypePtr->genericity()`).